File: articles/mathematical-biology-logic-living-systems/r/logistic_growth.R
# Logistic growth workflow in R.

scenario_path <- file.path("data", "logistic_scenarios.csv")
if (!file.exists(scenario_path)) {
  scenario_path <- file.path("..", "data", "logistic_scenarios.csv")
}

logistic_growth <- function(time, N0, r, K) {
  K / (1 + ((K - N0) / N0) * exp(-r * time))
}

scenarios <- read.csv(scenario_path)

rows <- list()

for (i in seq_len(nrow(scenarios))) {
  s <- scenarios[i, ]
  time <- seq(0, s$time_end, by = s$dt)
  population <- logistic_growth(time, s$initial_population, s$growth_rate, s$carrying_capacity)

  rows[[i]] <- data.frame(
    scenario = s$scenario,
    final_population = tail(population, 1),
    fraction_of_capacity = tail(population, 1) / s$carrying_capacity,
    initial_doubling_time = log(2) / s$growth_rate
  )
}

summary_df <- do.call(rbind, rows)
print(round(summary_df, 5))

File: articles/differential-equations-population-physiological-modeling/python/logistic_growth.py
"""
Logistic growth scenarios.

Run:
    python python/logistic_growth.py
"""

from pathlib import Path

import pandas as pd

from differential_equations_core import LogisticParameters, simulate_logistic


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "logistic_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)
    rows = []

    for _, row in scenarios.iterrows():
        params = LogisticParameters(row["N0"], row["r"], row["K"], row["dt"], row["t_end"])
        trajectory = simulate_logistic(params)

        rows.append(
            {
                "scenario": row["scenario"],
                "final_population": trajectory["population"].iloc[-1],
                "fraction_of_capacity": trajectory["population"].iloc[-1] / row["K"],
                "max_population": trajectory["population"].max(),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
